ramp4-pcar4/ramp4-pcar4
# ESRI Chunk Load

### Related Item(s)

#2155 (should relatively conclude the RAMP side of this issue. Storyilnes / Respect bundler still a target for enhancements).

### Changes

- Wires up on-demand chunk loading for most ESRI libraries when using the `esDynamic` build, or importing from `npm` combined with having a respectful bundler.
- ESRI Geometry and Symbology classes remain pre-bundled, as introducing async on all of them would not have been worth the tradeoff.
- Improved error handling of poorly configured basemaps.
- Found and squashed small bug around loading server name from MIL Raster children.
- Removed an incorrect piece of doc in `permanentFilteredQuery` section.

### Notes

The main starter chunk dropped from 2.7 mb to 1.3 mb :rocket:.

**Breaking Changes:**

The following Map API calls now return a Promise that resolves with the original synchronous value. This affects both the primary map (`MapAPI`) and the overview map (`OverviewMapAPI`).
- `createMap()`
- `reloadMap()`
- `findBasemap()`
- `setBasemap()`
- `recoverBasemap()`

The class constructor for `Basemap` will no longer initialize the internal ESRI basemap class. Use the new static method `Basemap.create(config)`, which returns a Promise that resolves in the initialized basemap.

Typical sites will not be impacted by these API changes. Using `createInstance()` works as it always has. But any fancy page code or custom fixtures directly hitting these endpoints will need to adjust.

### QA Testing

Please test against `main` branch https://ramp4-pcar4.github.io/ramp4-pcar4/main/demos/enhanced-samples.html

### Testing

General integrity tests:

- General map and layer loading
- Changing basemaps (with and without map schema change)
- Language change (use an Enhanced sample to ensure its not a singular config and true map reload)
- Classic Sample 33 - basemap failover (will take 10 seconds or so but basemap should appear)
- Classic Sample 44 - delayed map load

Developer network activity test:

1. Open Enhanced Sample 1
1. Mash `F12`, open `Network` tab, clear all the activity
1. Paste the below script in the console. You should see something named `@arcgis_core_layers_MapImageLayer.js` get downloaded, since this is the first MIL on this sample.

```js
const inputConfig = {
    id: 'EcoGeo',
    layerType: 'esri-map-image',
    url: 'https://section917.canadacentral.cloudapp.azure.com/arcgis/rest/services/TestData/EcoAction/MapServer/',
    sublayers: [{ index: 6 }, { index: 8 }]
};
const lay = debugInstance.geo.layer.createLayer(inputConfig);
debugInstance.geo.map.addLayer(lay).then(()=> { 
  debugInstance.event.emit('user/layeradded', lay);
});
```


- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/ramp4-pcar4/ramp4-pcar4/2529)
